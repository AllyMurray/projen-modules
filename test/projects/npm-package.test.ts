import { synthSnapshot } from 'projen/lib/util/synth';
import { TypeScriptNpmPackage } from '../../src/projects/npm-package';

describe('TypeScriptNpmPackage', () => {
  it('synthesizes', () => {
    const project = new TypeScriptNpmPackage({
      name: 'test',
      authorName: 'Ally Murray',
      defaultReleaseBranch: 'main',
    });

    const output = synthSnapshot(project);

    expect(output).toMatchSnapshot();
  });
});
