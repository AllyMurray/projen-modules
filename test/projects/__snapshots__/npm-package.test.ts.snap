// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScriptNpmPackage synthesizes 1`] = `
{
  ".eslintrc.json": {
    "env": {
      "jest": true,
      "node": true,
    },
    "extends": [
      "plugin:import/typescript",
      "prettier",
      "plugin:prettier/recommended",
    ],
    "ignorePatterns": [
      "*.js",
      "*.d.ts",
      "node_modules/",
      "*.generated.ts",
      "coverage",
      "scripts/create-package-json.ts",
      "!.projenrc.ts",
      "!projenrc/**/*.ts",
    ],
    "overrides": [
      {
        "files": [
          "*.test.ts",
        ],
        "rules": {
          "dot-notation": "off",
          "import/no-extraneous-dependencies": [
            "error",
            {
              "devDependencies": true,
            },
          ],
        },
      },
      {
        "files": [
          ".projenrc.ts",
        ],
        "rules": {
          "@typescript-eslint/no-require-imports": "off",
          "import/no-extraneous-dependencies": "off",
        },
      },
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
      "ecmaVersion": 2018,
      "project": "./tsconfig.dev.json",
      "sourceType": "module",
    },
    "plugins": [
      "@typescript-eslint",
      "import",
      "prettier",
    ],
    "root": true,
    "rules": {
      "@typescript-eslint/member-ordering": [
        "error",
        {
          "default": [
            "public-static-field",
            "public-static-method",
            "protected-static-field",
            "protected-static-method",
            "private-static-field",
            "private-static-method",
            "field",
            "constructor",
            "method",
          ],
        },
      ],
      "@typescript-eslint/no-floating-promises": [
        "error",
      ],
      "@typescript-eslint/no-require-imports": [
        "error",
      ],
      "@typescript-eslint/no-shadow": [
        "error",
      ],
      "@typescript-eslint/return-await": [
        "error",
      ],
      "dot-notation": [
        "error",
      ],
      "import/no-extraneous-dependencies": [
        "error",
        {
          "devDependencies": [
            "**/test/**",
            "**/build-tools/**",
            ".projenrc.ts",
            "projenrc/**/*.ts",
          ],
          "optionalDependencies": false,
          "peerDependencies": true,
        },
      ],
      "import/no-unresolved": [
        "error",
      ],
      "import/order": [
        "warn",
        {
          "alphabetize": {
            "caseInsensitive": true,
            "order": "asc",
          },
          "groups": [
            "builtin",
            "external",
          ],
        },
      ],
      "key-spacing": [
        "error",
      ],
      "no-bitwise": [
        "error",
      ],
      "no-duplicate-imports": [
        "error",
      ],
      "no-multiple-empty-lines": [
        "error",
      ],
      "no-return-await": [
        "off",
      ],
      "no-shadow": [
        "off",
      ],
      "no-trailing-spaces": [
        "error",
      ],
      "prettier/prettier": [
        "error",
      ],
    },
    "settings": {
      "import/parsers": {
        "@typescript-eslint/parser": [
          ".ts",
          ".tsx",
        ],
      },
      "import/resolver": {
        "node": {},
        "typescript": {
          "alwaysTryTypes": true,
          "project": "./tsconfig.dev.json",
        },
      },
    },
  },
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

/.eslintrc.json linguist-generated
/.gitattributes linguist-generated
/.github/pull_request_template.md linguist-generated
/.github/workflows/build.yml linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/release.yml linguist-generated
/.github/workflows/upgrade-main.yml linguist-generated
/.gitignore linguist-generated
/.mergify.yml linguist-generated
/.npmignore linguist-generated
/.npmrc linguist-generated
/.prettierignore linguist-generated
/.prettierrc.json linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/LICENSE linguist-generated
/package.json linguist-generated
/pnpm-lock.yaml linguist-generated
/scripts/create-package-json.ts linguist-generated
/tsconfig.build.cjs.json linguist-generated
/tsconfig.build.esm.json linguist-generated
/tsconfig.dev.json linguist-generated
/tsconfig.json linguist-generated",
  ".github/pull_request_template.md": "Fixes #",
  ".github/workflows/build.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: "8"
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: build
        run: npx projen build
      - name: Find mutations
        id: self_mutation
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT
      - name: Upload patch
        if: steps.self_mutation.outputs.self_mutation_happened
        uses: actions/upload-artifact@v3
        with:
          name: .repo.patch
          path: .repo.patch
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
          cat .repo.patch
          exit 1
  self-mutation:
    needs: build
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: always() && needs.build.outputs.self_mutation_happened && !(github.event.pull_request.head.repo.full_name != github.repository)
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Download patch
        uses: actions/download-artifact@v3
        with:
          name: .repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/.repo.patch ] && git apply \${{ runner.temp }}/.repo.patch || echo "Empty patch. Skipping."'
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Push changes
        env:
          PULL_REQUEST_REF: \${{ github.event.pull_request.head.ref }}
        run: |-
          git add .
          git commit -s -m "chore: self mutation"
          git push origin HEAD:$PULL_REQUEST_REF
",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v5.0.2
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
          githubBaseUrl: \${{ github.api_url }}
",
  ".github/workflows/release.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: release
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  release:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      latest_commit: \${{ steps.git_remote.outputs.latest_commit }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          fetch-depth: 0
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: "8"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: release
        run: npx projen release
      - name: Check for new commits
        id: git_remote
        run: echo "latest_commit=$(git ls-remote origin -h \${{ github.ref }} | cut -f1)" >> $GITHUB_OUTPUT
      - name: Backup artifact permissions
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        run: cd dist && getfacl -R . > permissions-backup.acl
        continue-on-error: true
      - name: Upload artifact
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        uses: actions/upload-artifact@v3
        with:
          name: build-artifact
          path: dist
  release_github:
    name: Publish to GitHub Releases
    needs: release
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: needs.release.outputs.latest_commit == github.sha
    steps:
      - uses: actions/setup-node@v3
        with:
          node-version: 16.x
      - name: Download build artifacts
        uses: actions/download-artifact@v3
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Release
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
          GITHUB_REPOSITORY: \${{ github.repository }}
          GITHUB_REF: \${{ github.ref }}
        run: errout=$(mktemp); gh release create $(cat dist/releasetag.txt) -R $GITHUB_REPOSITORY -F dist/changelog.md -t $(cat dist/releasetag.txt) --target $GITHUB_REF 2> $errout && true; exitcode=$?; if [ $exitcode -ne 0 ] && ! grep -q "Release.tag_name already exists" $errout; then cat $errout; exit $exitcode; fi
  release_npm:
    name: Publish to npm
    needs: release
    runs-on: ubuntu-latest
    permissions:
      contents: read
    if: needs.release.outputs.latest_commit == github.sha
    steps:
      - uses: actions/setup-node@v3
        with:
          node-version: 16.x
      - name: Download build artifacts
        uses: actions/download-artifact@v3
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Release
        env:
          NPM_DIST_TAG: latest
          NPM_REGISTRY: registry.npmjs.org
          NPM_TOKEN: \${{ secrets.NPM_TOKEN }}
        run: npx -p publib@latest publib-npm
",
  ".github/workflows/upgrade-main.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: upgrade-main
on:
  workflow_dispatch: {}
  schedule:
    - cron: 0 0 * * *
jobs:
  upgrade:
    name: Upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: read
    outputs:
      patch_created: \${{ steps.create_patch.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: main
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: "8"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Upgrade dependencies
        run: npx projen upgrade
      - name: Find mutations
        id: create_patch
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo "patch_created=true" >> $GITHUB_OUTPUT
      - name: Upload patch
        if: steps.create_patch.outputs.patch_created
        uses: actions/upload-artifact@v3
        with:
          name: .repo.patch
          path: .repo.patch
  pr:
    name: Create Pull Request
    needs: upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: read
    if: \${{ needs.upgrade.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: main
      - name: Download patch
        uses: actions/download-artifact@v3
        with:
          name: .repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/.repo.patch ] && git apply \${{ runner.temp }}/.repo.patch || echo "Empty patch. Skipping."'
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Create Pull Request
        id: create-pr
        uses: peter-evans/create-pull-request@v4
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          commit-message: |-
            chore(deps): upgrade dependencies

            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: \${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the "upgrade-main" workflow*
          branch: github-actions/upgrade-main
          title: "chore(deps): upgrade dependencies"
          body: |-
            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: \${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the "upgrade-main" workflow*
          author: github-actions <github-actions@github.com>
          committer: github-actions <github-actions@github.com>
          signoff: true
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/package.json
!/LICENSE
!/.npmignore
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
pids
*.pid
*.seed
*.pid.lock
lib-cov
coverage
*.lcov
.nyc_output
build/Release
node_modules/
jspm_packages/
*.tsbuildinfo
.eslintcache
*.tgz
.yarn-integrity
.cache
.DS_Store
*yalc*
test-reports
!/.projenrc.js
!/.github/workflows/build.yml
/dist/changelog.md
/dist/version.txt
!/.github/workflows/release.yml
!/.mergify.yml
!/.github/workflows/upgrade-main.yml
!/.github/pull_request_template.md
!/.prettierignore
!/.prettierrc.json
!/.npmrc
!/test/
!/tsconfig.json
!/tsconfig.dev.json
!/src/
/lib
/dist/
!/.eslintrc.json
!/scripts/create-package-json.ts
!/tsconfig.build.esm.json
!/tsconfig.build.cjs.json
",
  ".mergify.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

queue_rules:
  - name: default
    update_method: merge
    conditions:
      - "#approved-reviews-by>=1"
      - -label~=(do-not-merge)
      - status-success=build
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        method: squash
        name: default
        commit_message_template: |-
          {{ title }} (#{{ number }})

          {{ body }}
    conditions:
      - "#approved-reviews-by>=1"
      - -label~=(do-not-merge)
      - status-success=build
",
  ".npmignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
/.projen/
permissions-backup.acl
/dist/changelog.md
/dist/version.txt
/.mergify.yml
/test/
/tsconfig.dev.json
/src/
!/lib/
!/lib/**/*.js
!/lib/**/*.d.ts
dist
/tsconfig.json
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.eslintrc.json
/tsconfig.build.esm.json
/tsconfig.build.cjs.json
",
  ".npmrc": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

resolution-mode=highest
fund=false
no-bin-links=true
",
  ".prettierignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
scripts/create-package-json.ts
",
  ".prettierrc.json": {
    "overrides": [],
    "singleQuote": true,
  },
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "dependencies": [
      {
        "name": "@types/node",
        "type": "build",
        "version": "^16",
      },
      {
        "name": "@typescript-eslint/eslint-plugin",
        "type": "build",
        "version": "^6",
      },
      {
        "name": "@typescript-eslint/parser",
        "type": "build",
        "version": "^6",
      },
      {
        "name": "commander",
        "type": "build",
      },
      {
        "name": "eslint-config-prettier",
        "type": "build",
      },
      {
        "name": "eslint-import-resolver-node",
        "type": "build",
      },
      {
        "name": "eslint-import-resolver-typescript",
        "type": "build",
      },
      {
        "name": "eslint-plugin-import",
        "type": "build",
      },
      {
        "name": "eslint-plugin-prettier",
        "type": "build",
      },
      {
        "name": "eslint",
        "type": "build",
        "version": "^8",
      },
      {
        "name": "npm-check-updates",
        "type": "build",
        "version": "^16",
      },
      {
        "name": "prettier",
        "type": "build",
      },
      {
        "name": "projen",
        "type": "build",
      },
      {
        "name": "standard-version",
        "type": "build",
        "version": "^9",
      },
      {
        "name": "ts-node",
        "type": "build",
      },
      {
        "name": "typescript",
        "type": "build",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "files": [
      ".eslintrc.json",
      ".gitattributes",
      ".github/pull_request_template.md",
      ".github/workflows/build.yml",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/release.yml",
      ".github/workflows/upgrade-main.yml",
      ".gitignore",
      ".mergify.yml",
      ".npmignore",
      ".npmrc",
      ".prettierignore",
      ".prettierrc.json",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "LICENSE",
      "scripts/create-package-json.ts",
      "tsconfig.build.cjs.json",
      "tsconfig.build.esm.json",
      "tsconfig.dev.json",
      "tsconfig.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "env": {
      "PATH": "$(pnpm -c exec "node --print process.env.PATH")",
    },
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "bump": {
        "condition": "! git log --oneline -1 | grep -q "chore(release):"",
        "description": "Bumps version based on latest git tag and generates a changelog entry",
        "env": {
          "BUMPFILE": "dist/version.txt",
          "CHANGELOG": "dist/changelog.md",
          "OUTFILE": "package.json",
          "RELEASETAG": "dist/releasetag.txt",
          "RELEASE_TAG_PREFIX": "",
        },
        "name": "bump",
        "steps": [
          {
            "builtin": "release/bump-version",
          },
        ],
      },
      "clobber": {
        "condition": "git diff --exit-code > /dev/null",
        "description": "hard resets to HEAD of origin and cleans the local repo",
        "env": {
          "BRANCH": "$(git branch --show-current)",
        },
        "name": "clobber",
        "steps": [
          {
            "exec": "git checkout -b scratch",
            "name": "save current HEAD in "scratch" branch",
          },
          {
            "exec": "git checkout $BRANCH",
          },
          {
            "exec": "git fetch origin",
            "name": "fetch latest changes from origin",
          },
          {
            "exec": "git reset --hard origin/$BRANCH",
            "name": "hard reset to origin commit",
          },
          {
            "exec": "git clean -fdx",
            "name": "clean all untracked files",
          },
          {
            "say": "ready to rock! (unpushed commits are under the "scratch" branch)",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
        "steps": [
          {
            "exec": "tsc --project tsconfig.build.esm.json",
          },
          {
            "exec": "ts-node ./scripts/create-package-json.ts --moduleType="esm" --packageJsonType=module",
          },
          {
            "exec": "tsc --project tsconfig.build.cjs.json",
          },
          {
            "exec": "ts-node ./scripts/create-package-json.ts --moduleType="cjs" --packageJsonType=commonjs",
          },
        ],
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "eslint": {
        "description": "Runs eslint against the codebase",
        "name": "eslint",
        "steps": [
          {
            "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools projenrc .projenrc.ts",
          },
        ],
      },
      "install": {
        "description": "Install project dependencies and update lockfile (non-frozen)",
        "name": "install",
        "steps": [
          {
            "exec": "pnpm i --no-frozen-lockfile",
          },
        ],
      },
      "install:ci": {
        "description": "Install project dependencies using frozen lockfile",
        "name": "install:ci",
        "steps": [
          {
            "exec": "pnpm i --frozen-lockfile",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
        "steps": [
          {
            "exec": "mkdir -p dist/js",
          },
          {
            "exec": "mv $(npm pack) dist/js/",
          },
        ],
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "post-upgrade": {
        "description": "Runs after upgrading dependencies",
        "name": "post-upgrade",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "release": {
        "description": "Prepare a release from "main" branch",
        "env": {
          "RELEASE": "true",
        },
        "name": "release",
        "steps": [
          {
            "exec": "rm -fr dist",
          },
          {
            "spawn": "bump",
          },
          {
            "spawn": "build",
          },
          {
            "spawn": "unbump",
          },
          {
            "exec": "git diff --ignore-space-at-eol --exit-code",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
        "steps": [
          {
            "exec": "vitest --dir=src",
            "receiveArgs": true,
          },
          {
            "spawn": "eslint",
          },
        ],
      },
      "unbump": {
        "description": "Restores version to 0.0.0",
        "env": {
          "BUMPFILE": "dist/version.txt",
          "CHANGELOG": "dist/changelog.md",
          "OUTFILE": "package.json",
          "RELEASETAG": "dist/releasetag.txt",
          "RELEASE_TAG_PREFIX": "",
        },
        "name": "unbump",
        "steps": [
          {
            "builtin": "release/reset-version",
          },
        ],
      },
      "upgrade": {
        "description": "upgrade dependencies",
        "env": {
          "CI": "0",
        },
        "name": "upgrade",
        "steps": [
          {
            "exec": "pnpm update npm-check-updates",
          },
          {
            "exec": "npm-check-updates --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,commander,eslint-config-prettier,eslint-import-resolver-node,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,npm-check-updates,prettier,projen,standard-version,ts-node,typescript",
          },
          {
            "exec": "pnpm i --no-frozen-lockfile",
          },
          {
            "exec": "pnpm update @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser commander eslint-config-prettier eslint-import-resolver-node eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint npm-check-updates prettier projen standard-version ts-node typescript",
          },
          {
            "exec": "npx projen",
          },
          {
            "spawn": "post-upgrade",
          },
        ],
      },
      "watch": {
        "description": "Watch & compile in the background",
        "name": "watch",
        "steps": [
          {
            "exec": "tsc --build -w",
          },
        ],
      },
    },
  },
  "LICENSE": "Copyright (c) 2023 Ally Murray

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
",
  "README.md": "# replace this",
  "package.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "author": {
      "email": "allymurray88@gmail.com",
      "name": "Ally Murray",
      "organization": false,
    },
    "devDependencies": {
      "@types/node": "^16",
      "@typescript-eslint/eslint-plugin": "^6",
      "@typescript-eslint/parser": "^6",
      "commander": "*",
      "eslint": "^8",
      "eslint-config-prettier": "*",
      "eslint-import-resolver-node": "*",
      "eslint-import-resolver-typescript": "*",
      "eslint-plugin-import": "*",
      "eslint-plugin-prettier": "*",
      "npm-check-updates": "^16",
      "prettier": "*",
      "projen": "*",
      "standard-version": "^9",
      "ts-node": "*",
      "typescript": "*",
    },
    "exports": {
      ".": {
        "import": {
          "default": "./lib/esm/index.js",
          "types": "./lib/esm/index.d.ts",
        },
        "require": {
          "default": "./lib/cjs/index.js",
          "types": "./lib/cjs/index.d.ts",
        },
      },
    },
    "license": "MIT",
    "main": "lib/cjs/index.js",
    "module": "lib/esm/index.js",
    "name": "test",
    "pnpm": {},
    "scripts": {
      "build": "npx projen build",
      "bump": "npx projen bump",
      "clobber": "npx projen clobber",
      "compile": "npx projen compile",
      "default": "npx projen default",
      "eject": "npx projen eject",
      "eslint": "npx projen eslint",
      "package": "npx projen package",
      "post-compile": "npx projen post-compile",
      "post-upgrade": "npx projen post-upgrade",
      "pre-compile": "npx projen pre-compile",
      "projen": "npx projen",
      "release": "npx projen release",
      "test": "npx projen test",
      "unbump": "npx projen unbump",
      "upgrade": "npx projen upgrade",
      "watch": "npx projen watch",
    },
    "type": "module",
    "types": "lib/cjs/index.d.ts",
    "version": "0.0.0",
  },
  "scripts/create-package-json.ts": "// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
/**
 * Creates a package.json for the given build type
 */
import fs from 'fs';
import path from 'path';
import { program } from 'commander';

program
  .option('--moduleType <string>', 'The module type, cjs or esm')
  .option(
    '--packageJsonType <string>',
    'The package.json type, either commonjs or module'
  );
program.parse();

const options = program.opts();

const type = options.packageJsonType;

fs.writeFileSync(
  path.resolve('lib', options.moduleType, 'package.json'),
  JSON.stringify(
    {
      type,
    },
    null,
    2
  )
);",
  "tsconfig.build.cjs.json": {
    "compilerOptions": {
      "module": "commonjs",
      "outDir": "lib/cjs",
      "target": "es2015",
    },
    "exclude": [
      ".projenrc.ts",
      "test",
      "src/**/*.spec.ts",
      "src/**/*.test.ts",
    ],
    "extends": "./tsconfig.json",
    "include": [
      "src/**/*.ts",
    ],
  },
  "tsconfig.build.esm.json": {
    "compilerOptions": {
      "outDir": "lib/esm",
    },
    "exclude": [
      ".projenrc.ts",
      "test",
      "src/**/*.spec.ts",
      "src/**/*.test.ts",
    ],
    "extends": "./tsconfig.json",
    "include": [
      "src/**/*.ts",
    ],
  },
  "tsconfig.dev.json": {
    "compilerOptions": {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": [
        "es2019",
      ],
      "module": "Node16",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "resolveJsonModule": true,
      "skipLibCheck": true,
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2020",
    },
    "exclude": [
      "node_modules",
    ],
    "include": [
      ".projenrc.js",
      "src/**/*.ts",
      "test/**/*.ts",
      ".projenrc.ts",
      "projenrc/**/*.ts",
    ],
  },
  "tsconfig.json": {
    "compilerOptions": {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": [
        "es2019",
      ],
      "module": "Node16",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "outDir": "lib",
      "resolveJsonModule": true,
      "rootDir": "src",
      "skipLibCheck": true,
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2020",
    },
    "exclude": [],
    "include": [
      "src/**/*.ts",
    ],
  },
}
`;
